internal/command/console/console.go

```diff
@@ -0,0 +1,349 @@
+		long  = "Run a console in a new or existing machine. The console command is\n" +
+			"specified by the `console_command` configuration field. By default, a\n" +
+			"new machine is created by default using the app's most recently deployed\n" +
+			"image. An existing machine can be used instead with --machine."
+	)
+	cmd := command.New(usage, short, long, runConsole, command.RequireSession, command.RequireAppName)
+
+	cmd.Args = cobra.NoArgs
+	flag.Add(
+		cmd,
+		flag.App(),
+		flag.AppConfig(),
+		flag.Int{
+			Name:        "cpus",
+			Description: "How many (shared) CPUs to give the new machine",
+		},
+		flag.String{
+			Name:        "machine",
+			Description: "Run the console in the existing machine with the specified ID",
+		},
+		flag.Int{
+			Name:        "memory",
+			Description: "How much memory (in MB) to give the new machine",
+		},
+		flag.Bool{
+			Name:        "select",
+			Shorthand:   "s",
+			Description: "Select the machine on which to execute the console from a list",
+			Default:     false,
+		},
+		flag.String{
+			Name:        "user",
+			Shorthand:   "u",
+			Description: "Unix username to connect as",
+			Default:     ssh.DefaultSshUsername,
+		},
+	)
+
+	return cmd
+}
+
+func runConsole(ctx context.Context) error {
+	var (
+		io        = iostreams.FromContext(ctx)
+		colorize  = io.ColorScheme()
+		appName   = appconfig.NameFromContext(ctx)
+		apiClient = client.FromContext(ctx).API()
+	)
+
+	app, err := apiClient.GetAppCompact(ctx, appName)
+	if err != nil {
+		return fmt.Errorf("failed to get app: %w", err)
+	}
+
+	if app.PlatformVersion != "machines" {
+		return errors.New("console is only supported for the machines platform")
+	}
+
+	flapsClient, err := flaps.New(ctx, app)
+	if err != nil {
+		return fmt.Errorf("failed to create flaps client: %w", err)
+	}
+	ctx = flaps.NewContext(ctx, flapsClient)
+
+	appConfig := appconfig.ConfigFromContext(ctx)
+	if appConfig == nil {
+		appConfig, err = appconfig.FromRemoteApp(ctx, appName)
+		if err != nil {
+			return fmt.Errorf("failed to fetch app config from backend: %w", err)
+		}
+	}
+
+	if err, extraInfo := appConfig.ValidateForMachinesPlatform(ctx); err != nil {
+		fmt.Fprintln(io.ErrOut, extraInfo)
+		return err
+	}
+
+	machine, ephemeral, err := selectMachine(ctx, app, appConfig)
+	if err != nil {
+		return err
+	}
+
+	if ephemeral {
+		defer func() {
+			const stopTimeout = 5 * time.Second
+
+			stopCtx, cancel := context.WithTimeout(context.Background(), stopTimeout)
+			defer cancel()
+
+			stopInput := api.StopMachineInput{
+				ID:      machine.ID,
+				Timeout: api.Duration{Duration: stopTimeout},
+			}
+			if err := flapsClient.Stop(stopCtx, stopInput, ""); err != nil {
+				terminal.Warnf("Failed to stop ephemeral runner machine: %v\n", err)
+				terminal.Warn("You may need to destroy it manually (`fly machine destroy`).")
+				return
+			}
+
+			fmt.Fprintf(io.Out, "Waiting for ephemeral runner machine %s to be destroyed ...", colorize.Bold(machine.ID))
```

Comment:
Just tried the command, it was amazing but this wait at the end felt weird. It blocked my terminal with a "long" pause before it becomes usable again.